add initial docs for windows client testing

jessfraz opened the pull request:
ping @ahmetalpbalkan

jessfraz committed:
add initial docs for windows client testing

Docker-DCO-1.1-Signed-off-by: Jessie Frazelle <princess@docker.com> (github: jfrazelle)

Docker-DCO-1.1-Signed-off-by: Jessie Frazelle <jess@docker.com> (github: jfrazelle)

ahmetalpbalkan commented on docs/sources/project/windows-client-test-server.md:
```diff
@@ -0,0 +1,139 @@
+- TDM-GCC
+- MinGW (tar and xz)
+- Go with linux/amd64 cross-compiling
+- Docker.exe
```
I'm not sure if we need this one.

jfrazelle replied:
its nice for testing your remote host is set up correctly but yeah we can make optional

ahmetalpbalkan replied:
Yeah it sounds like a hard requirement. I don't think we need to explain where to grab docker as well.

ahmetalpbalkan replied:
I mean people are more likely to run into problems like version mismatches etc. I don't think this step is necessary as they can get a docker.exe with `hack/make.sh binary` simply.

moxiegirl replied:
When I write these kinda things, I always include the versions I used in writing the procedure. This makes troubleshooting for the reader and anyone in support easier.  The baseline to look at it is the env.

SvenDowideit replied:
+1 to versions - otherwise users will use the old one they might already have installed, and not know that they need to update

ahmetalpbalkan commented on docs/sources/project/windows-client-test-server.md:
```diff
@@ -0,0 +1,139 @@
+Things you will need:
+
+- Windows Server/Desktop Machine
+- Java
```
not sure where we need Java either. Jenkins slave stuff?

jfrazelle replied:
my b lol

ahmetalpbalkan commented on docs/sources/project/windows-client-test-server.md:
```diff
@@ -0,0 +1,139 @@
+
+Let's try it.
+
+1. Open GitBash.
```
"Git Bash" to be precise ^_^

ahmetalpbalkan commented on docs/sources/project/windows-client-test-server.md:
```diff
@@ -0,0 +1,139 @@
+2. Enter the docker source directory we set up `cd
+   /c/gopath/src/github.com/docker/docker`.
+
+3. Run `project/make.sh binary test-integration-cli` to make the binary and run
```
I think we must include `TESTFLAGS='-test.v' ` here as well as quite many tests are skipped on windows and skips are not shown clearly without -test.v

tiborvass replied:
Also adding `-test.run $TESTNAME` with an example would be great to explain how they can run just one test.

ahmetalpbalkan replied:
@tiborvass not sure about the verboseness level here. Contributing docs already have it and maybe Go programmers can infer they can add -test.run as -test.v is already there?

ahmetalpbalkan commented on docs/sources/project/windows-client-test-server.md:
```diff
@@ -0,0 +1,139 @@
+   your path. Go to Control Panel > System and Security > System, then click the Advanced
+   System Settings link in the sidebar.
+
+4. In the Advanced tab, click Enviornment Variables. Now scroll to the Path
```
I'm not big fan of polluting people's global environment variable settings. They're really hard to debug and contaminating. 

Why don't we just advise them to run the following in git bash

```sh
cd /c/gopath/src/github.com/docker/docker
export GOPATH=/c/gopath/src/github.com/docker/docker/vendor:/c/gopath
export DOCKER_CLIENTONLY=1
export DOCKER_TEST_HOST=tcp://localhost:2376
TESTFLAGS='-test.v' project/make.sh binary test-integration-cli
```

thaJeztah replied:
Also s/Enviornment/Environment/

SvenDowideit replied:
+1 to not polluting the global env where possible.

ahmetalpbalkan commented on docs/sources/project/windows-client-test-server.md:
```diff
@@ -0,0 +1,139 @@
+    `GOPATH` variable in the screen from step 3 as
+    `C:\gopath;C:\gopath\src\github.com\docker\docker\vendor`.
+
+## Install docker.exe
```
I'm -1 for having this section :( And even more -1 to adding this to PATH.

ahmetalpbalkan commented:
Can't leave comments on images but I guess we need to crop those to contain only the windows screen?

jfrazelle commented:
leeeeenux, I will put gimp in a container lol

moxiegirl commented:
@jfrazelle this is essentially what I've already got written and tested in the contributor's guide for Mac/Unbuntu  --- I was going to test today against Windows 7 so this is a huge boost for me.   Um, I'll keep tracking the comments on this; I would restructure a bit for ease of use.   So, I'll test against my restructure which is basically going to be (minus markdown eff ups) this:


-----

page_title: Set up a Windows server for client testing
page_description: How to set up a server to test Docker Windows client
page_keywords: development, inception, container, image Dockerfile, dependencies, Go, artifacts, windows


# Windows Client test server

In this section, you learn how to setup a Windows machine for building and
testing the Docker Windows client. 

## Prerequisites

Before you start this procedure, make sure you have a Windows server or desktop
machine with the following installed:

- Java
- MSysGit
- TDM-GCC
- MinGW (tar and xz)
- Go with linux/amd64 cross-compiling
- Docker.exe

### Installing Java

From your Windows desktop:

1. Open your browser and go to [java.com](http://java.com).

3. Click the "Free Java Download" button.

		The system prompts you to accept the license agreement.

4. Click "Agree and Start Free Download." 

4. On the notification bar, click "Run."


### Installing MSysGit

1. Open your browser and go to [msysgit.github.io](https://msysgit.github.io/).

2. Click the button to download "MSysGit."

### Installing GCC

1. Open your browser and go to
   [tdm-gcc.tdragon.net/download](http://tdm-gcc.tdragon.net/download).

2. Click the lastest version of the package.

3. Download and install the package.

### Installing MinGW (tar and xz)

1. Download the latest MinGW from
   [SourceForge](http://sourceforge.net/projects/mingw/).
   
   The system installs the MinGW Installation Manager.

2. Open the MinGW Installation Manager.

3. Select `tar` and `xz` as shown below:
    
    ![windows-mingw](/project/images/windows-mingw.png)
    
4. Select "Installation > Apply Changes".

		The system installs the selected packages.



## Make sure gcc is in your path

1. Go to "Control Panel > System and Security > System."

2. Click the "Advanced System Settings" link in the sidebar.

3. In the "Advanced" tab, click "Environment Variables."

4. In the "System" variables list, scroll to the "Path variable.

5. Double-click the "Path" variable to see what it includes.

	![windows-env-vars](/project/images/windows-env-vars.png)
	
6. Look the version in the `C:\TDM-GCC64` path.

		You must have at least 64 to cross compile Go.
		
7. Exit it out of the dialog being sure to have your new settings.
    

## Installing Golang with linux/amd64 cross-compiling

1. Open your browser  to [golang.org/dl](http://golang.org/dl/).

2. Download and install the latest `.msi` installer.
   .
   Wait for the installation to complete.
   
3. Open `gitbash.`

4. Change the directory where `go` was installed. 
 
 			`cd /c/Go/src`
 		
 		The location may be different on your machine.

5. Run `cmd.exe`. 

	 You should now be in the `cmd.exe` prompt.

6. Set the following environment variables.
        
        set GOOS=linux
        set GOARCH=amd64
        
7. Run the `make` command.

        make.bat

    This will allow for cross compiling for linux/amd64. If the `make` fails with
    a cannot find `gcc` message, go back to step 1 and ensure you have `gcc` in your path.
    
    

## Download the Docker source and Windows client

	1. Create a folder `C:\gopath\src\github.com\docker\docker`.
		
	2. Clone the Docker repository into it. 
	
	3. Create a directory `C:\bin`.

	4. Download the latest master build of the Windows client from
   [master.dockerproject.com](https://master.dockerproject.com).

	5. Place this file in the directory `C:\bin`.


## Update the PATH in your environment

	1. Go to "Control Panel > System and Security > System."
		
	2. Click the "Advanced System Settings" link in the sidebar.

	3. In the "Advanced" tab, click "Enviornment Variables."
		
	4. In the "System" variables list, scroll to the "PATH" variable. 
	
	5. Edit the "PATH" variable and add the following to the end of it.
		
			  C:\gopath;C:\gopath\src\github.com\docker\docker\vendor

   6. Add `C:\bin` to the end of the variable if it is not already
   there.

    ![windows-env-vars](/project/images/windows-env-vars.png)

5. Add a `DOCKER_HOST` variable.

	 This connects to your Boot2Docker instance with `docker` installed. 

6. Add a `DOCKER_TEST_HOST` to save value as `DOCKER_HOST`.
	 
7. Add `DOCKER_CLIENTONLY` variable.

			DOCKER_CLIENTONLY=1
			  
	9. Exit it out of the dialog being sure to have your new settings.
	
10. Open a GitBash shell.

		You'll use this shell to test your setup.

11. Run the following command to test your environment.

			docker run hello-world
			
## Building and testing the Windows Docker Client

You should now have a windows server and remote host all set up for building and testing docker. 

Let's try it.

1. Open GitBash.

2. Change to the `docker` source directory 

		cd /c/gopath/src/github.com/docker/docker`.

3. Run `project/make.sh binary test-integration-cli` command.

	  This builds the `docker` binary the integration tests.

If you make any changes just run these commands again.

ahmetalpbalkan commented:
@moxiegirl umm what's the diff between the doc you pasted and Jess wrote?

SvenDowideit commented on docs/mkdocs.yml:
```diff
@@ -175,6 +175,7 @@ pages:
 - ['project/who-written-for.md', 'Project', 'README first']
 - ['project/set-up-prereqs.md', 'Project', 'Set up prerequisites'] 
 - ['project/set-up-dev-env.md', 'Project', 'Work with a development container'] 
+- ['project/windows-client-test-server.md', 'Project', 'Set up a Windows server for client testing'] 
```
I _think_ using the word `server` in this context is misleading - `Windows system` will then also include normal windows versions too.

SvenDowideit commented on docs/sources/project/windows-client-test-server.md:
```diff
@@ -0,0 +1,139 @@
+page_keywords: development, inception, container, image Dockerfile, dependencies, Go, artifacts, windows
+
+
+# Windows Client test server
```
replace `server` with `system`

ahmetalpbalkan replied:
also the filename, I'd recommend changing this to "Testing Windows Docker Client"

SvenDowideit replied:
how about `Building and testing the Windows Docker client`? otherwise, someone looking to find our how to build Docker on windows might not read this one.

SvenDowideit commented:
nice - I wonder if we should add some info wrt Boot2Docker users - as they have some of the pre-req's

LK4D4 commented:
ping @ahmetalpbalkan @jfrazelle @moxiegirl @SvenDowideit 
This PR needs your attention.

ahmetalpbalkan commented:
LGTM.

thaJeztah commented:
@ahmetalpbalkan @jfrazelle  I'll do some cropping and cleaning up with the images if you want, can send a PR to your branch. (Had it on my todo list).

Will try later tonight, otherwise tomorrow if there's no hurry

jfrazelle commented:
wow sorry completely forgot about this one I will update and install gimp to cleanup screenshots

jfrazelle commented:
oh @thaJeztah that would be great because image editing on leeenux is not ideal :)

thaJeztah commented:
@jfrazelle right, will do. Not sure if I'll make it in today, might be tomorrow.

jfrazelle commented:
no worries take your time :)

jfrazelle commented:
made updates based off comments, can we maybe get this in then @thaJeztah can make a follow up PR when he has time to clean up images, since he was so kind and offered :angel:

thaJeztah committed:
docs: crop and optimize images

Cropped and optimized images.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

Docker-DCO-1.1-Signed-off-by: Jessie Frazelle <princess@docker.com> (github: jfrazelle)

jfrazelle commented:
ping @SvenDowideit @moxiegirl :)

jfrazelle commented:
cherry picked @thaJeztah commit here :dancer:

moxiegirl commented:
@jfrazelle I'm thinking this belongs in the installation folder...It doesn't yet quite fit into the project dir  --- That dir is for the contributors guide.  I was going to update this to fit it into that flow but the wave of distribution hit me. What do you think about moving it to install?

jfrazelle commented:
sgtm!

On Fri, Apr 3, 2015 at 7:32 PM, moxiegirl <notifications@github.com> wrote:

> @jfrazelle <https://github.com/jfrazelle> I'm thinking this belongs in
> the installation folder...It doesn't yet quite fit into the project dir ---
> That dir is for the contributors guide. I was going to update this to fit
> it into that flow but the wave of distribution hit me. What do you think
> about moving it to install?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11296#issuecomment-89398971>.
>

ahmetalpbalkan commented:
> thinking this belongs in the installation folder

@moxiegirl I don't think so, this is _far_ too complicated for a user. There are probably only 3-4 people who need to use this document at this point and I don't see this getting picked up very popularly. Belongs to Contributing Guide IMO.

jfrazelle commented:
ah eff i I just moved it :(

jfrazelle commented:
ok decide and then ping me

jfrazelle commented:
whats the consensus

ahmetalpbalkan commented:
Waiting on input from @moxiegirl. IMO Installation category had nothing to do with this doc.

moxiegirl commented:
Thanks for the move @jfrazelle . LGTM.

@ahmetalpbalkan Let's merge it into install we can move it later. It really doesn't belong in the project contributors guide.  We could put it into articles but honestly it won't get the hits it would get in the installation section.  Your choice.

moxiegirl commented:
@ahmetalpbalkan The project contributors guide is structured in a very specific way by me for the typical contributor. The first step is described  here http://docs.docker.com/project/software-required/ and says to install Docker.  Most of this document is about getting docker up and running on server 2012

If I understand correctly, Jess tested this install on a server edition of windows. It seems unlikely most of our contributors are using a server edition of Windows.  It seems like an advance use case.  My goal is to get this into the docs&mdash; in install as the majority of the content is there.  I intend to revisit it in light of the contributor guide later.

Hopefully you are willing to compromise as, like I said, the piece is likely to get on the order of 55K users in install.  Fully 75% of the people that hit project leave.

ahmetalpbalkan commented:
@moxiegirl my argument was we prepared this document in order to not to forget setup steps for the windows test environment. I don't think the remaining 55K readers need it nor they will probably do anything useful with this. :)

moxiegirl commented:
I should also add I have a very specific Google Analytics function running on the flow of the sequence in the project.  I watch the set very carefully because of the conversion we are seeking to use.  I'm really not comfortable Jaming this into the set.

crosbymichael commented:
This is a very technical document and maybe the location of the document depending on it's contents is not the correct place.  You can tell it is a little different just by looking at the header diff.

```patch
 - ['installation/ubuntulinux.md', 'Installation', 'Ubuntu']
 - ['installation/mac.md', 'Installation', 'Mac OS X']
 - ['installation/windows.md', 'Installation', 'Microsoft Windows']
+- ['installation/testing-windows-docker-client.md', 'Installation', 'Building and testing the Windows Docker client']
 - ['installation/amazon.md', 'Installation', 'Amazon EC2']
 - ['installation/archlinux.md', 'Installation', 'Arch Linux']
 - ['installation/binaries.md', 'Installation', 'Binaries']
```

Pick out the line that does not look like it belongs ;)  

I think this is really good content either for the contributor guide or even just a `README.md` for the windows code in the client.

moxiegirl commented:
I'm cool with it as README.  As I said, I'll revisit it after this release for inclusion in the contributor's guide.

jfrazelle commented:
so where does the readme go? im fine w whatever :)

crosbymichael commented:
My initial guess would be in the client code

jfrazelle commented:
ok i moved to in the integration-cli directory....

jfrazelle commented:
can we please get this in now?

moxiegirl commented:
@jfrazelle I'm gonna carry it in another PR to fit into the guide. Thanks for being flexible.
